cpp/src/arrow/util/parsing.h

```diff
@@ -374,13 +374,25 @@ class StringConverter<TimestampType> {
     // - "YYYY-MM-DD[ T]hh:mmZ"
     // - "YYYY-MM-DD[ T]hh:mm:ss"
     // - "YYYY-MM-DD[ T]hh:mm:ssZ"
+    // - "MM/DD/YYYY"
```

Comment:
As [discussed in the JIRA](https://issues.apache.org/jira/browse/ARROW-8111?focusedCommentId=17058819&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-17058819), this format is too locale specific for an unconditional inference like this (as it only makes sense for dates originating in the US). Please remove it

Reply:
Agreed. This will have to be explicitly enabled by user-supplied options. I suggest that we make the option general enough to allow for other date formats.

Reply:
I see, any suggestions where can I put this option? I think option like `datetimeLocale=US/EU` would be good enough to solve that ambiguity.

Reply:
or we can add options `dayfirst` and `yearfirst` like in pandas

Reply:
I think instead of adding options like this the better solution would be to provide support for a strtptime format. See https://issues.apache.org/jira/browse/ARROW-3739

Reply:
My gut feeling is that we should center efforts on the API user supplying strptime-like format strings like "%m/%d/%Y". For now you can special case the one you have implemented until a more general parser can be implemented.

At some point we might also need to add a functional API where the user passes an opaque parsing function.

Reply:
I see, but we still need a place where we can specify the format.

Reply:
I'd say it can be an optional argument to the constructor of `StringConverter<TimestampType>`.

Reply:
`StringConverter<TimestampType>` is used here

https://github.com/apache/arrow/blob/master/cpp/src/arrow/csv/converter.cc#L393

So two options:

* Introduce an alternative code path for user-supplied format (versus using the ISO-8601 converter) -- this is probably what I would do
* Add options to `StringConverter<TimestampType>`. @pitrou may have a preference since he has worked on this code

Reply:
I could start implementing strptime based parser if you propose API for that :)

Reply:
@fexolm You could start by extending `StringConverter<TimestampType>` with a strptime format string. Once that's done if it winds up being too much functionality to add to that header, you should be able extract it to a separate class/header straightforwardly.

Reply:
I think you should want to add a field like `timestamp_format` to `ConvertOptions`

https://github.com/apache/arrow/blob/master/cpp/src/arrow/csv/options.h#L63

Reply:
If we wanted to augment timestamp type inference to support many different formats, then we could also add an API like

```
std::vector<std::string> timestamp_infer_formats = {"iso8601", "%m/%d/%Y"};
```

Others may have opinions about the public API

Reply:
Hmm... I think we should be careful to keep the common case as fast as it is. I'm not sure if that has to be through the use of different converters, or if the compiler will be smart enough to duplicate loops.

Reply:
> I think we should be careful to keep the common case as fast as it is. 

Yes, I agree. Probably each format parser should be a separate loop -- so if you just want to infer and parse iso8601 then it is equivalent to the current version of the code. Having an inner loop over formats sounds like a bad idea

Reply:
@pitrou @wesm @bkietz Folks, maybe we would better implement smth like that
`std::vector<TimestampParser> timestamp_parsers = {Iso8691Parser, StrptimeParser("%m/%d/%Y"), MyCustomParser };`
This would give us good extendability.
As for our purpose, we want fast parser, supporting multiple formats, so we would better write our own parser, without using multiple strptime calls.

Reply:
I think that would push a lot of complexity on the user. Rather than making them choose what parser to use just have them provide a timestamp format string and choose the best implementation for them. i.e. in python pseudo-code:

```python
ISO8601_FORMAT = '%Y-%m-%dT%H:%M:%S'

registry: Dict[str, TimestampParser] = {}
registry[ISO8601_FORMAT] = Iso8601Parser


def get_parser(fmt: str) -> TimestampParser:
    return registry.get(fmt, partial(StrptimeParser, fmt=fmt))
```

Reply:
@dhirschfeld that way we can not use multiple parsers to one column. E.g in the mortgage dataset they have 'mm/dd/yyyy' and 'mm/yyyy' in the same column. 
Anyway, the code you provided could be the internal implementation of StrptimeParser if mapping formats to implementation would give performance speedup, but I don't really think so.
The code you provided could be rewritten to  `timestamp_parsers = { StrptimeParser("%Y-%m-%dT%H:%M:%S") };`  without pushing comlexity, but keeping opportunity to add aditional parsers here.

Reply:
I've updated pr with my latest proposal, PTAL.

Reply:
@dhirschfeld we're talking about C++ here, so we can craft an API in Python or R to hide these kinds of details.